### PR TITLE
test: Add basic MTE tests for InventoryManager

### PR DESCRIPTION
**Pull request by skaldarnar**
Adds tests for the `InventoryManager` using the [ModuleTestingEnvironment](https://github.com/Terasology/ModuleTestingEnvironment).
This makes use of https://github.com/Terasology/ModuleTestingEnvironment/pull/14 as well as some experimental features in the engine.

✅ Depends on https://github.com/Terasology/ModuleTestingEnvironment/pull/14
✅ Depends on https://github.com/MovingBlocks/Terasology/pull/3977

### CODE REVIEW

**Comment by skaldarnar** [comment]
While the results don't look that good yet, at least running the tests in an acceptable amount if time is great:
image

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+        entityManager = hostContext.get(EntityManager.class);
+        inventoryManager = hostContext.get(InventoryManager.class);
+        blockManager = hostContext.get(BlockManager.class);
+
+        inventory = entityManager.create(new InventoryComponent(3));
+    }
+
+    @ParameterizedTest
+    @ValueSource(bytes = {0, -1})
```
Comment (jdrueckert) [inline src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:77]: why `bytes` parameter but `int` function input?

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        EntityRef blockItem = getDirtBlockItem(entityManager, blockManager, amount);
+        inventoryManager.giveItem(inventory, EntityRef.NULL, blockItem);
+
+        final List<EntityRef> filledSlots = getFilledSlots(inventory);
+        Assert.assertEquals("Exactly one slot should be filled", 1, filledSlots.size());
+        filledSlots.forEach(item -> {
+            Assert.assertEquals("Slot should contain exactly the added block", blockItem, item);
+
+            final int maxStackSize = blockItem.getComponent(ItemComponent.class).maxStackSize;
+            Assert.assertEquals("Stack count should be " + amount + " or maximum stack size",
```
Comment (jdrueckert) [inline src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:98]: I think in this test, the stack count should be exactly `amount` as you don't have any excess...

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            Assert.assertEquals("Slot should contain exactly the added block", blockItem, item);
+
+            final int maxStackSize = blockItem.getComponent(ItemComponent.class).maxStackSize;
+            Assert.assertEquals("Stack count should be capped at maximum stack size",
+                    Math.min(amount, maxStackSize), item.getComponent(ItemComponent.class).stackCount);
+        });
+    }
+
+    @ParameterizedTest
+    @ValueSource(bytes = {1, 2})
```
Comment (jdrueckert) [inline src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:121]: any reason you use `bytes` in some tests and `ints` in others?

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    public void giveItem_nonStackableItem(byte amount) {
```
Comment (jdrueckert) [inline src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:122]: why isn't there a `giveItem_zeroItem` test?

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+            final int maxStackSize = originalItem.getComponent(ItemComponent.class).maxStackSize;
+            Assert.assertEquals("Unexpected stack count",
+                    1, slot.getComponent(ItemComponent.class).stackCount);
+
+        });
+    }
+
+    @ParameterizedTest
+    @ValueSource(bytes = {1, 2, 99})
```
Comment (jdrueckert) [inline src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:146]: I think the default stack size for stackable items is 16 🤔

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class InventoryManagerTest {
+
+    static final String URI_DIRT = "CoreAssets:Dirt";
+
+    private static ModuleTestingEnvironment context;
```
Comment (jdrueckert) [inline src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:41]: In #12 you call this `mte`. In the spirit of consistency, consider changing either to mirror the other :)

Comment (skaldarnar) [reply src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:41]: I think `context` might be misleading (we have a couple of _contexts_ around here), so I'd go for `mte` (or `testEnvironment`).

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java (hunk 1 above)
Comment (skaldarnar) [reply src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:77]: This is a mistake here. 
In general, the methods around blocks accept integer values, whereas for items we have to set the quantity manually as byte - no way to get an int into that...

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    @ParameterizedTest
+    @ValueSource(bytes = {0, -1})
+    public void giveItem_zeroBlocks(int amount) {
+        EntityRef blockItem = getDirtBlockItem(entityManager, blockManager, amount);
+        inventoryManager.giveItem(inventory, EntityRef.NULL, blockItem);
+
+        final List<EntityRef> filledSlots = getFilledSlots(inventory);
+        Assert.assertEquals("No slots should be filled", 0, filledSlots.size());
```
Comment (DarkWeird) [inline src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:83]: Use junit5's `Assertion` instead junit4's `Assert`. 
Most methods similar, except parameters order: message is last parameter.

File: src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+    @ParameterizedTest
+    @ValueSource(ints = {100, 128, 256})
+    public void giveItem_blockExceedingStacksize(int amount) {
+        EntityRef blockItem = getDirtBlockItem(entityManager, blockManager, amount);
+        inventoryManager.giveItem(inventory, EntityRef.NULL, blockItem);
+
+        final List<EntityRef> filledSlots = getFilledSlots(inventory);
+        Assert.assertEquals("Exactly one slot should be filled", 1, filledSlots.size());
+        filledSlots.forEach(item -> {
```
Comment (DarkWeird) [inline src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:111]: We junit 5's `Assetions:
```
Assertions.assertAll(
    filledSlots.stream()
                    .map((item)-> ()-> { // Lambda of Executable - Consumer with throws exception.
                        /// Your assertions.
   });
```
Also, you can use faltMap, if you want.

Comment (skaldarnar) [reply src/test/java/org/terasology/logic/inventory/InventoryManagerTest.java:103]: Good point, but became a bit lengthy ....

```java
        Assertions.assertAll(
                filledSlots.stream().flatMap(item ->
                        Stream.of(
                                () -> Assertions.assertEquals(blockItem, item,
                                        "Slot should contain exactly the added block"),
                                () -> Assertions.assertEquals(amount, item.getComponent(ItemComponent.class).stackCount,
                                        "Unexpected slot count!")
                        )
                )
        );
```

I'm pretty sure I'll use that at some point, though ^^

**Pull request merged**